### DISCUSSION TITLE
Allow admin to generate auth user sessions

### DISCUSSION THREAD

**Discussion by ioucyf** [supabase/supabase#11854] [Feature Requests]
I'm trying to use the WebAuthn API with supabase as one of the primary sign-in options on my web client; since it's not yet officially supported by supabase, I'm writing a custom API to handle the registration and verification of the user's authenticator.

However, while using supabase's admin client with the service role key on the server API, I have no way of generating a valid session, for that specific user after confirming that the user's authenticator key is in fact valid, to return it to the client; to be set as the new session via: (`supabase.auth.setSession(<received new session>)`). 

... or is there? I don't know...

Please make it possible to do so.

**Comment by ioucyf** [comment]
Is it possible to do create a valid user session by generating the JWTs myself using the `secret key` of the project's API?

**Reply by ruggi99** [reply]
Hi @ofeenee,
yes it is possible to do.
I suggest you to use `jwt.io` site.
Take a valid token and put it in the textarea, then fill in the secret, change `iat` field (aka Issued At) and you should have done.
Actually I can't confirm my steps are all right, but I'm close

**Reply by ioucyf** [reply]
Can I , as Admin, set the session on the supabase client for a user to be valid and authenticated using a JWT I generated with the projects secret key?

**Reply by ruggi99** [reply]
I think you can do it with a bunch of lines of client-side code.
https://github.com/supabase/gotrue-js/blob/013afaeb45e5d964a8cca467b748a3daa557f22c/src/GoTrueClient.ts#L650

**Reply by roo12312** [reply]
Were you able to create session?

**Reply by kirill-writer** [reply]
hey @ioucyf , I was able to generate a user session (by email) on the BE and send it to the FE for further authentication.

BE part
```
    const { data: linkData, error: linkError } =
      await supabaseAdmin.auth.admin.generateLink({
        type: 'magiclink',
        email: email,
      });

    if (linkError || !linkData?.properties?.hashed_token) {
      return NextResponse.json(
        { error: 'Failed to generate session token' },
        { status: 500 },
      );
    }

    const { data: otpData, error: otpError } =
      await supabaseAdmin.auth.verifyOtp({
        token_hash: linkData.properties.hashed_token,
        type: 'magiclink',
      });

    const { access_token, refresh_token } = otpData.session;

    return NextResponse.json({
      access_token,
      refresh_token,
      user,
    });
```

FE part
```
      const res = await fetch('/api/auth/supabase-token');
      if (!res.ok) {
        console.log('No valid token found or token expired');
        return false;
      }

      const { access_token, refresh_token } = await res.json();

      const { data, error } = await supabase.auth.setSession({
        access_token,
        refresh_token,
      });

// user is authenticated here
```

**Comment by anribras** [comment]
Have you done it ? @ioucyf  I faced same issue. Seems no way unless you modify "auth" module  which looks pretty complex.
